### PR TITLE
On-Board check-catalyst workflow to use GitHub Large Runners

### PR DESCRIPTION
**Pull request by rashidnhm**
**Context:** 
When high volume of jobs are being queued on CI. It is possible that it will take a while before a job is picked up. To get around this, we have provisioned an organization runner group of GitHub Large Runners. Since these are paid runners, we retain the ability to scale this up to a value suitable for us during periods of high usage.

**Description of the Change:**
A new label `urgent` has been introduced to this repo. Whenever a pull request has this label attached, all the jobs within check-catalyst.yml will swap over to use the new Large runner group described above.

This PR also bumps the version of actions/checkout and actions/cache being used. This was done to deal with the deprecation of node16 within github actions.

**Benefits:**
Lowered queue times for urgent CI workloads to get picked up.

**Possible Drawbacks:**
We can still see jobs being queued for a while if enough PRs have the `urgent` label at once. To get around this, only add it to PRs that are truly urgent. Also, we can increase the scaling factor of the group if needed.

**Related GitHub Issues:**
None. [sc-61791](https://app.shortcut.com/xanaduai/story/61791/update-catalyst-ci-to-use-large-runners)

[sc-61184]: fix for cache issues

### CODE REVIEW

File: .github/workflows/check-catalyst.yaml [hunk 1]
```diff
@@ -139,57 +147,57 @@ jobs:
...
       if: steps.cache-mhlo.outputs.cache-hit != 'true'
-      uses: actions/cache@v3
+      uses: actions/cache@v4
       with:
         path: mlir/llvm-project
         key: ${{ runner.os }}-llvm-${{ needs.constants.outputs.llvm_version }}-default-source
-        enableCrossOsArchive: True
-        fail-on-cache-miss: True
+        enableCrossOsArchive: true
+        fail-on-cache-miss: true
```
Comment (erick-xanadu) [inline .github/workflows/check-catalyst.yaml:191]: Is `True` is not the correct keyword for toml? :sweat_smile:

File: .github/workflows/determine-workflow-runner.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  LARGE_RUNNER_GROUP_NAME: pl-4-core-large-runner
+
+jobs:
+  determine_workflow_runner:
+    runs-on: >-
+      ${{
+        (
+          github.event_name == 'pull_request'
+          && contains(github.event.pull_request.labels.*.name, 'urgent')
+        ) && 'pl-4-core-large-runner' || 'ubuntu-latest'
```
Comment (dime10) [inline .github/workflows/determine-workflow-runner.yml:25]: I don't quite follow this sequence, what do the two values at the end do?

**Review by dime10 — COMMENTED** [review]
Thanks @rashidnhm, can we include the linux wheels in this?

File: .github/workflows/check-catalyst.yaml (hunk 1 above)
Comment (dime10) [reply .github/workflows/check-catalyst.yaml:191]: This is `yaml`, but also here is the github docs for it: https://docs.github.com/en/actions/learn-github-actions/expressions#literals

File: .github/workflows/determine-workflow-runner.yml (hunk 2 above)
Comment (rashidnhm) [reply .github/workflows/determine-workflow-runner.yml:25]: It is an if statement:
```
if (github.event_name == 'pull_request' && contains(..) {
  runs-on: large runner
} else {
  runs-on: ubuntu-latest
}
```

Comment (dime10) [reply .github/workflows/determine-workflow-runner.yml:25]: Oh I see, a bit obscure but it works 😅

**Comment by dime10** [comment]
It would be nice to upgrade invoked actions across all our workflows (like what you did in `check-catalyst-build`), or would you like to do that in a follow up?

File: .github/workflows/build-wheel-linux-x86_64.yaml [hunk 3]
```diff
@@ -317,7 +325,7 @@ jobs:
         cmake -S mlir -B quantum-build -G Ninja \
               -DCMAKE_BUILD_TYPE=Release \
               -DLLVM_ENABLE_ASSERTIONS=ON \
-              -DQUANTUM_ENABLE_BINDINGS_PYTHON=ON \
+              -DQUAfromJsonNTUM_ENABLE_BINDINGS_PYTHON=ON \
```
Comment (dime10) [inline .github/workflows/build-wheel-linux-x86_64.yaml:328]:
```suggestion
              -DQUANTUM_ENABLE_BINDINGS_PYTHON=ON \
```

**Review by dime10 — APPROVED** [review]

**Comment by rashidnhm** [comment]
> It would be nice to upgrade invoked actions across all our workflows (like what you did in `check-catalyst.yml`), or would you like to do that in a follow up?

It shouldn't be that much effort to upgrade I'll add it in within this pr

**Review by erick-xanadu — APPROVED** [review]

**Review by erick-xanadu — CHANGES_REQUESTED** [review]
I think we need to revert the checkout action to version 3: https://github.com/actions/checkout/issues/1487 https://github.com/actions/checkout/issues/1474

**Comment by erick-xanadu** [comment]
I think that we can switch to checkout action v4 when we switch to another container. But manylinux_2014 is likely unsupported. See these screenshots and this link:
https://github.com/tarantool/tarantool/commit/686bd701ea2f3e23e927b98467a6fec8c8198e21

Screenshot from 2024-04-25 16-21-22
Screenshot from 2024-04-25 16-21-44

**Review by erick-xanadu — APPROVED** [review]

**Pull request merged**